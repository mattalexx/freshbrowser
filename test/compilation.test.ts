import {fork} from 'child_process'
import {mkdtemp} from 'fs/promises'
import {tmpdir} from 'os'
import {join} from 'path'
import ts, {ExitStatus} from 'typescript'
import {ParsedCommandLine} from 'typescript/lib/tsserverlibrary'
import {childClosed, ChildClosedPayload, ChildCloseOptions, replacePaths, rootDir} from './support/utils'

interface TestCase {
    tsconfig: string
}

const tscExe = join(rootDir, 'node_modules', 'typescript', 'bin', 'tsc')

const TIMEOUT = 10000

const tsProjects: Array<TestCase> = [
    // Scripts
    {tsconfig: 'scripts'},
    // Source
    {tsconfig: 'src'},
    // Tests
    {tsconfig: 'test'},
]

describe('TypeScript project config', () => {
    it.each(tsProjects)('file should parse and resolve correctly: $tsconfig', async ({tsconfig}) => {
        const {stdout} = await tsc(['--project', abs(tsconfig), '--showConfig'])
        const stdoutNomalized = replacePaths(stdout as string)
        const json = JSON.parse(stdoutNomalized)
        const config: ParsedCommandLine = ts.parseJsonConfigFileContent(json, ts.sys, abs(tsconfig))
        expect(config.raw).toMatchSnapshot()
    })

    it.each(tsProjects as TestCase[])('should compile and build without errors: $tsconfig', async ({tsconfig}) => {
        await tsc(['--project', abs(tsconfig), '--noEmit'])
    }, TIMEOUT)

    it.each(tsProjects)('should compile and build without errors: $tsconfig', async ({tsconfig}) => {
        await tsc(['--build', abs(tsconfig)])
    }, TIMEOUT)

    it('the root project should build without errors: ./tsconfig.json', async () => {
        await tsc(['--build', rootDir])
    }, TIMEOUT)
})

async function tsc(cmdArgs: string[], closeOpts?: ChildCloseOptions): Promise<ChildClosedPayload> {
    const tmp: Promise<string> = mkdtemp(join(tmpdir(), 'darkreader'))
    const forkOpts = {silent: true, ourCwd: tmp}
    const child = fork(tscExe, cmdArgs, forkOpts)
    const payload = await childClosed(child, closeOpts)
    const error = child.exitCode === 0 ? undefined : payload.stdout;
    expect(error).toBeUndefined()
    expect(child.exitCode).toBe(ExitStatus.Success)
    return payload
}

function abs(relPath: string) {
    return join(rootDir, relPath, 'tsconfig.json')
}
