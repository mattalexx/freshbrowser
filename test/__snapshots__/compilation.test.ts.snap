// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript project config file should parse and resolve correctly: scripts 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "outDir": "../out",
    "rootDir": "..",
    "typeRoots": Array [
      "<PROJECT_DIR>/node_modules/@types",
      "<PROJECT_DIR>/src/types",
    ],
  },
  "files": Array [
    "./launch-ppt.ts",
  ],
  "include": Array [
    "**/*.ts",
    "**/*.tsx",
    "**/*.js",
    "**/*.jsx",
  ],
  "references": Array [
    Object {
      "path": "../src",
    },
  ],
}
`;

exports[`TypeScript project config file should parse and resolve correctly: src 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "composite": true,
    "declaration": true,
    "declarationMap": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "lib": Array [
      "es2021",
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "outDir": "../out",
    "rootDir": "..",
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "target": "es2021",
    "types": Array [
      "node",
      "puppeteer",
    ],
  },
  "exclude": Array [
    "../out",
  ],
  "files": Array [
    "./data.ts",
    "./extension.ts",
    "./prelaunch.ts",
    "./project.ts",
    "./types/env-paths/index.d.ts",
    "./types/unzip-crx-3/index.d.ts",
  ],
  "include": Array [
    "**/*.ts",
    "**/*.tsx",
    "**/*.js",
    "**/*.jsx",
  ],
}
`;

exports[`TypeScript project config file should parse and resolve correctly: test 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "allowJs": true,
    "allowSyntheticDefaultImports": true,
    "composite": true,
    "declaration": true,
    "declarationMap": true,
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "lib": Array [
      "es2021",
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "outDir": "../out",
    "rootDir": "..",
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "target": "es2021",
    "types": Array [
      "node",
      "jest",
      "puppeteer",
    ],
  },
  "exclude": Array [
    "../out",
  ],
  "files": Array [
    "./compilation.test.ts",
    "./extension.test.ts",
    "./project.test.ts",
    "./tsconf.tests.ts",
    "./support/utils.ts",
  ],
  "include": Array [
    "**/*.ts",
    "**/*.tsx",
    "**/*.js",
    "**/*.jsx",
  ],
  "references": Array [
    Object {
      "path": "../src",
    },
  ],
}
`;
